sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -586,9 +584,139 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
+      "org.apache.spark.sql.types.StringType")
+    if (leftIsString && rightIsString) {
+      Seq(left.dataType, right.dataType)
+    } else {
+      Seq(StringType, StringType)
+    }
+  }
+
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val lCollationID = left.dataType.asInstanceOf[StringType].collationId
+    val rCollationID = right.dataType.asInstanceOf[StringType].collationId
+    // TODO: ensure that lCollationID and rCollationID are equal (compatible)
+    val collationID = if (lCollationID > rCollationID) lCollationID else rCollationID
+    val collatorInfo = CollatorFactory.getInfoForId(collationID)
+    val collator = collatorInfo.collator
+    collator match {
+      // UTF8String collation
+      case null =>
+        collationID match {
+          // UCS_BASIC
+          case 0 =>
+            l.contains(r)
+          // UCS_BASIC_LCASE
+          case 1 =>
+            // scalastyle:off caselocale
+            l.toLowerCase.contains(r.toLowerCase)
+          // scalastyle:on caselocale
+          // Other special collations without ICU collator
+          case _ =>
+            var found = false
+            for (i <- 0 to l.numBytes() - r.numBytes() if !found) {
+              if (collatorInfo.comparator.compare(l.substring(i, i + r.numBytes()), r) == 0) {
+                found = true
+              }
+            }
+            found
+        }
+      // String ICU collation
+      case _ =>
+        val lStr: String = l.toString
+        val rStr: String = r.toString
+        var found = false
+        for (i <- 0 to lStr.length - rStr.length if !found) {
+          if (collator.compare(lStr.substring(i, i + rStr.length), rStr) == 0) {
+            found = true
+          }
+        }
+        found
+    }
+  }
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    defineCodeGen(ctx, ev, (c1, c2) => s"($c1).contains($c2)")
+    val lCollationID = left.dataType.asInstanceOf[StringType].collationId
+    val rCollationID = right.dataType.asInstanceOf[StringType].collationId
+    // TODO: ensure that lCollationID and rCollationID are equal (compatible)
+    val collationID = if (lCollationID > rCollationID) lCollationID else rCollationID
+    val collatorFactory = ctx.addReferenceObj("collatorFactory", CollatorFactory.getInstance())
+    val collatorInfo = CollatorFactory.getInfoForId(collationID)
+    val collator = collatorInfo.collator
+    val leftGen = if (left.dataType.sameType(StringType)) {
+      left.genCode(ctx)
+    } else {
+      left.asInstanceOf[Collate].inputString.genCode(ctx)
+    }
+    val rightGen = if (right.dataType.sameType(StringType)) {
+      right.genCode(ctx)
+    } else {
+      right.asInstanceOf[Collate].inputString.genCode(ctx)
+    }
+    collator match {
+      case null =>
+        collationID match {
+          case 0 =>
+            // UCS_BASIC
+            defineCodeGen(ctx, ev, (c1, c2) => s"($c1).contains($c2)")
```

Comment:
Note: this is the original implementation before adding collation support (works only for UTF8String as is).

(see above for corresponding Scala code)

Reply:
Note: We keep using this for UCS_BASIC in the new [PR](https://github.com/apache/spark/pull/45216/files)